File: apps/kira-studio/frontend/src/views/console/planParsers/postgres.ts
import { isOverThreshold, maxEstimatedRows, pushWideScanIssue, rollupIssues } from '../planIssues';
import type { PlanNode, QueryPlan, ScanEstimate } from '../planModel';

// P18 (v1.1) D13/D15/D16: Postgres 18's `EXPLAIN (FORMAT JSON, COSTS TRUE, ...)` returns one row,
// one column QUERY PLAN, whose text is a JSON array of one `{ Plan: {...} }` — F11. Only fields
// available *without* ANALYZE are read here; §7.3's own check is that Actual Rows/Actual Total
// Time/Rows Removed by Filter/Sort Method/Shared Hit Blocks never appear below — those require
// ANALYZE, which this phase never issues (§0.3).
interface RawNode {
  'Node Type'?: string;
  'Relation Name'?: string;
  Alias?: string;
  'Index Name'?: string;
  'Startup Cost'?: number;
  'Total Cost'?: number;
  'Plan Rows'?: number;
  'Plan Width'?: number;
  Filter?: string;
  'Index Cond'?: string;
  'Hash Cond'?: string;
  'Join Type'?: string;
  Strategy?: string;
  'Sort Key'?: string[];
  'Group Key'?: string[];
  'Parallel Aware'?: boolean;
  Disabled?: boolean;
  Plans?: RawNode[];
}

// D14: the node types a full/partial *read* actually happens at — the ones whose own `Plan Rows`
// D14's estimatedRowsRead is the max over.
const SCAN_NODE_TYPES = new Set([
  'Seq Scan',
  'Index Scan',
  'Index Only Scan',
  'Bitmap Heap Scan',
  'CTE Scan',
  'Function Scan',
  'Foreign Scan',
]);

// Every raw field this parser reads into a typed slot — the rest fall through to `metrics`
// verbatim (D16), so a field Postgres adds later still shows up rather than vanishing silently.
const TYPED_KEYS = new Set([
  'Node Type',
  'Relation Name',
  'Startup Cost',
  'Total Cost',
  'Plan Rows',
  'Filter',
  'Index Cond',
  'Hash Cond',
  'Plans',
]);

function metricsFrom(raw: RawNode): Array<{ label: string; value: string }> {
  const out: Array<{ label: string; value: string }> = [];
  for (const [key, value] of Object.entries(raw)) {
    if (TYPED_KEYS.has(key) || value === undefined) continue;
    out.push({ label: key, value: Array.isArray(value) ? value.join(', ') : String(value) });
  }
  return out;
}

function buildNode(raw: RawNode, thresholdRows: number, scans: ScanEstimate[]): PlanNode {
  const nodeType = raw['Node Type'] ?? 'Node';
  const relation = raw['Relation Name'];
  const label = relation
    ? `${nodeType} on ${raw.Alias && raw.Alias !== relation ? `${relation} ${raw.Alias}` : relation}`
    : nodeType;

  const node: PlanNode = {
    label,
    relation,
    detail: raw.Filter ?? raw['Index Cond'] ?? raw['Hash Cond'],
    estimatedRows: raw['Plan Rows'],
    cost:
      raw['Total Cost'] !== undefined
        ? { total: raw['Total Cost'], startup: raw['Startup Cost'] }
        : undefined,
    metrics: metricsFrom(raw),
    issues: [],
    children: (raw.Plans ?? []).map((child) => buildNode(child, thresholdRows, scans)),
  };

  // D15: full scan with a predicate — a Seq Scan that still carries a Filter means the planner
  // read every row and then discarded most of them client-side.
  if (nodeType === 'Seq Scan' && raw.Filter !== undefined) {
    node.issues.push({
      severity: 'warn',
      code: 'full-scan',
      message: `full table scan on "${relation ?? '?'}" with a filter — no index was used`,
    });
  }
  // D15 (info): index-only scan is a good sign, not a problem — the panel is not only bad news.
  if (nodeType === 'Index Only Scan') {
    node.issues.push({
      severity: 'info',
      code: 'index-only-scan',
      message: `"${relation ?? '?'}" is read entirely from its index — no heap access needed`,
    });
  }
  // D15: a nested loop whose inner side is wide re-executes that inner scan once per outer row.
  if (nodeType === 'Nested Loop') {
    const wideInner = node.children.find((c) => (c.estimatedRows ?? 0) >= 10_000);
    if (wideInner) {
      node.issues.push({
        severity: 'warn',
        code: 'nested-loop-wide-inner',
        message: `nested loop repeats a scan estimated at ${wideInner.estimatedRows?.toLocaleString()} rows for every outer row`,
      });
    }
  }

  if (SCAN_NODE_TYPES.has(nodeType) && node.estimatedRows !== undefined) {
    scans.push({ node, rows: node.estimatedRows });
  }
  return node;
}

export function parsePostgresPlan(rawText: string, thresholdRows: number): QueryPlan {
  const parsed = JSON.parse(rawText) as Array<{ Plan: RawNode }>;
  const scans: ScanEstimate[] = [];
  const root = buildNode(parsed[0]?.Plan ?? {}, thresholdRows, scans);
  for (const scan of scans) pushWideScanIssue(scan.node, scan.rows, thresholdRows);

  const estimatedRowsRead = maxEstimatedRows(scans);
  return {
    kind: 'postgres',
    root,
    estimatedRowsRead,
    nativeCost: root.cost ? { value: root.cost.total, unit: 'postgres-planner' } : undefined,
    issues: rollupIssues(root),
    overThreshold: isOverThreshold(estimatedRowsRead, thresholdRows),
    raw: rawText,
  };
}

File: apps/kira-studio/internal/queryplan/postgres.go
package queryplan

import (
	"encoding/json"
	"fmt"
)

// postgres.go ports planParsers/postgres.ts verbatim: Postgres's `EXPLAIN (FORMAT JSON, COSTS
// TRUE, ...)` returns one row, one column QUERY PLAN, whose text is a JSON array of one
// `{ Plan: {...} }`. Only fields available *without* ANALYZE are read — this app never issues
// ANALYZE (§1.1/§8.3).

type pgRawNode struct {
	NodeType     *string           `json:"Node Type"`
	RelationName *string           `json:"Relation Name"`
	Alias        *string           `json:"Alias"`
	StartupCost  *float64          `json:"Startup Cost"`
	TotalCost    *float64          `json:"Total Cost"`
	PlanRows     *float64          `json:"Plan Rows"`
	Filter       *string           `json:"Filter"`
	IndexCond    *string           `json:"Index Cond"`
	HashCond     *string           `json:"Hash Cond"`
	Plans        []json.RawMessage `json:"Plans"`
}

// pgScanNodeTypes are the node types a full/partial *read* actually happens at — the ones whose
// own Plan Rows is what estimatedRowsRead is the max over (D14).
var pgScanNodeTypes = map[string]bool{
	"Seq Scan": true, "Index Scan": true, "Index Only Scan": true, "Bitmap Heap Scan": true,
	"CTE Scan": true, "Function Scan": true, "Foreign Scan": true,
}

// pgTypedKeys are every raw field this parser reads into a typed slot — the rest fall through to
// Metrics verbatim, so a field Postgres adds later still shows up rather than vanishing silently.
var pgTypedKeys = map[string]bool{
	"Node Type": true, "Relation Name": true, "Startup Cost": true, "Total Cost": true,
	"Plan Rows": true, "Filter": true, "Index Cond": true, "Hash Cond": true, "Plans": true,
}

func pgBuildNode(raw json.RawMessage, thresholdRows int, scans *[]float64) Node {
	var typed pgRawNode
	_ = json.Unmarshal(raw, &typed)
	rawMap := decodeRawObject(raw)

	nodeType := "Node"
	if typed.NodeType != nil {
		nodeType = *typed.NodeType
	}
	label := nodeType
	if typed.RelationName != nil {
		relation := *typed.RelationName
		if typed.Alias != nil && *typed.Alias != relation {
			label = fmt.Sprintf("%s on %s %s", nodeType, relation, *typed.Alias)
		} else {
			label = fmt.Sprintf("%s on %s", nodeType, relation)
		}
	}

	var detail *string
	switch {
	case typed.Filter != nil:
		detail = typed.Filter
	case typed.IndexCond != nil:
		detail = typed.IndexCond
	case typed.HashCond != nil:
		detail = typed.HashCond
	}

	var cost *Cost
	if typed.TotalCost != nil {
		cost = &Cost{Total: *typed.TotalCost, Startup: typed.StartupCost}
	}

	children := make([]Node, 0, len(typed.Plans))
	for _, childRaw := range typed.Plans {
		children = append(children, pgBuildNode(childRaw, thresholdRows, scans))
	}

	node := Node{
		Label:         label,
		EstimatedRows: typed.PlanRows,
		Cost:          cost,
		Metrics:       collectMetrics(rawMap, pgTypedKeys),
		Issues:        []Issue{},
		Children:      children,
	}
	if typed.RelationName != nil {
		node.Relation = *typed.RelationName
	}
	if detail != nil {
		node.Detail = *detail
	}

	relationOrPlaceholder := "?"
	if typed.RelationName != nil {
		relationOrPlaceholder = *typed.RelationName
	}

	// D15: full scan with a predicate — a Seq Scan that still carries a Filter means the planner
	// read every row and then discarded most of them client-side.
	if nodeType == "Seq Scan" && typed.Filter != nil {
		node.Issues = append(node.Issues, Issue{
			Severity: "warn", Code: "full-scan",
			Message: fmt.Sprintf("full table scan on %q with a filter — no index was used", relationOrPlaceholder),
		})
	}
	// D15 (info): index-only scan is a good sign, not a problem.
	if nodeType == "Index Only Scan" {
		node.Issues = append(node.Issues, Issue{
			Severity: "info", Code: "index-only-scan",
			Message: fmt.Sprintf("%q is read entirely from its index — no heap access needed", relationOrPlaceholder),
		})
	}
	// D15: a nested loop whose inner side is wide re-executes that inner scan once per outer row.
	if nodeType == "Nested Loop" {
		for _, c := range node.Children {
			rows := 0.0
			if c.EstimatedRows != nil {
				rows = *c.EstimatedRows
			}
			if rows >= 10_000 {
				node.Issues = append(node.Issues, Issue{
					Severity: "warn", Code: "nested-loop-wide-inner",
					Message: fmt.Sprintf("nested loop repeats a scan estimated at %s rows for every outer row", commaFormat(*c.EstimatedRows)),
				})
				break
			}
		}
	}

	// D15/D14: the "wide scan" issue is pushed here, before node's value is copied into its
	// parent's Children slice (plan.go's own note on why this cannot be deferred to a later pass
	// the way the TypeScript port's own buildNode+pushWideScanIssue split does it).
	if pgScanNodeTypes[nodeType] && typed.PlanRows != nil {
		*scans = append(*scans, *typed.PlanRows)
		pushWideScanIssue(&node, *typed.PlanRows, thresholdRows)
	}
	return node
}

func parsePostgresPlan(rawText string, thresholdRows int) (Plan, error) {
	var parsed []struct {
		Plan json.RawMessage `json:"Plan"`
	}
	if err := json.Unmarshal([]byte(rawText), &parsed); err != nil {
		return Plan{}, fmt.Errorf("queryplan: parse postgres plan: %w", err)
	}
	var planRaw json.RawMessage
	if len(parsed) > 0 {
		planRaw = parsed[0].Plan
	} else {
		planRaw = json.RawMessage("{}")
	}

	scans := make([]float64, 0)
	root := pgBuildNode(planRaw, thresholdRows, &scans)

	estimatedRowsRead := maxEstimatedRows(scans)
	var nativeCost *NativeCost
	if root.Cost != nil {
		nativeCost = &NativeCost{Value: root.Cost.Total, Unit: CostUnitPostgresPlanner}
	}
	return Plan{
		Kind: "postgres", Root: root, EstimatedRowsRead: estimatedRowsRead, NativeCost: nativeCost,
		Issues: rollupIssues(&root), OverThreshold: isOverThreshold(estimatedRowsRead, thresholdRows),
		Raw: rawText,
	}, nil
}
